src/OpenTelemetry/Metrics/MeterProviderSdk.cs

```diff
@@ -99,33 +111,68 @@ public class MeterProviderSdk
 
         internal List<MetricProcessor> ExportProcessors { get; } = new List<MetricProcessor>();
 
-        internal void MeasurementsCompleted(Instrument instrument, object? state)
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static DateTimeOffset GetDateTimeOffset()
+        {
+            int tick = Environment.TickCount;
+            if (tick == MeterProviderSdk.lastTick)
+            {
+                return MeterProviderSdk.lastTimestamp;
+            }
+
+            var dt = DateTimeOffset.UtcNow;
+            MeterProviderSdk.lastTimestamp = dt;
+            MeterProviderSdk.lastTick = tick;
+
+            return dt;
+        }
+
+        internal void MeasurementsCompleted(Instrument instrument, object state)
         {
             Console.WriteLine($"Instrument {instrument.Meter.Name}:{instrument.Name} completed.");
         }
 
-        internal void MeasurementRecorded<T>(Instrument instrument, T value, ReadOnlySpan<KeyValuePair<string, object?>> tags, object? state)
-            where T : unmanaged
+        internal void MeasurementRecorded<T>(Instrument instrument, T value, ReadOnlySpan<KeyValuePair<string, object>> tagsRos, object state)
+            where T : struct
         {
-            // Run Pre Aggregator Processors
+            // Get Instrument State
 
-            var measurmentContext = new MeasurementItem(instrument, new DataPoint<T>(value, tags));
+            if (!(state is InstrumentState instrumentState))
```

Comment:
When would this condition occur? we expect the .NET API to provide us with the right `InstrumentState`, as we called `listener.EnableMeasurementEvents(instrument, instrumentState);`, right?

Reply:
Yes, you are correct. It's just defensive coding.

Reply:
Also, it is possible for user to call EnableMeasurment with something else. In which case, we have a deeper problem anyways.